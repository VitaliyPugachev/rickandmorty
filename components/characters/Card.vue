<script lang="ts" setup>
import type { CharacterModel } from '~/models/CharacterModel';

defineProps<{character: CharacterModel}>();

</script>

<template>
  <UCard class="character-card" @click="$router.push(`/characters/${character.id}`)">

    <template #header>
        <section class="character-card__image-wrapper">
            <img class="character-card__image" :src="character.image" :alt="character.name">
        </section>
    </template>

    <header class="character-card__header">
        <UBadge color="blue" :label="character.name"/>
    </header>

    <section class="character-card__info">
        <div class="character-card__info-row">
            <span class="character-card__label">
                Gender:
            </span>
            <span class="character-card__value">
                {{ character.gender }}
            </span>
        </div>

        <div class="character-card__info-row">
            <span class="character-card__label">
                Type:
            </span>
            <span class="character-card__value">
                {{ character.type.length > 25 ? `${character.type.substring(0, 25)}...`: character.type }}
            </span>
        </div>

        <div class="character-card__info-row">
            <span class="character-card__label">
                Location:
            </span>
            <span class="character-card__value">
                {{ 
                    character.location.name.length > 25 
                    ? `${character.location.name.substring(0, 25)}...`
                    : character.location.name 
                }}
            </span>
        </div>

        <div class="character-card__info-row">
            <span class="character-card__label">
                Status:
            </span>
            <span class="character-card__value">
                {{ character.status }}
            </span>
        </div>
    </section>
  </UCard>
</template>

<style lang="scss">
.character-card {
    cursor: pointer;

    &__header {
        display: flex;
        justify-content: center;
    }
}
</style>