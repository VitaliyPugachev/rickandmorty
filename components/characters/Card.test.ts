import Card from "./Card.vue";
import {it, expect} from "vitest";
import { mount } from '@vue/test-utils';
import type { CharacterModel } from "~/models/CharacterModel";

const DifferentCards: CharacterModel[] = [
  {
    created: "2017-11-04T18:48:46.250Z",
    episode: ["https://rickandmortyapi.com/api/episode/1", "https://rickandmortyapi.com/api/episode/2"],
    gender: "Male",
    id: 1,
    image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
    location: {name: "Citadel of Ricks", url: "https://rickandmortyapi.com/api/location/3"},
    name: "Rick Sanchez",
    origin: {name: "Earth (C-137)", url: "https://rickandmortyapi.com/api/location/1"},
    species: "Human",
    status: "Alive",
    type: "",
    url: "https://rickandmortyapi.com/api/character/1"
  },
  {
    created: "2017-11-04T18:48:46.250Z",
    episode: ["https://rickandmortyapi.com/api/episode/1", "https://rickandmortyapi.com/api/episode/2"],
    gender: "Female",
    id: 2,
    image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
    location: {name: "Earth", url: "https://rickandmortyapi.com/api/location/3"},
    name: "Summer Smith",
    origin: {name: "Earth (C-137)", url: "https://rickandmortyapi.com/api/location/1"},
    species: "Human",
    status: "Dead",
    type: "type_test",
    url: "https://rickandmortyapi.com/api/character/1"
  }
]

it.only.each(DifferentCards)('Card $name', (item) => {
  const wrapper = mount(Card, {
    slots: {
      default: 'label'
    },
    props: {
      character: item
    }
  });

  if (item.status === 'Alive') {
    expect(wrapper.html()).contains('Alive');
  } else {
    expect(wrapper.html()).not.contains('Alive');
  }

  expect(wrapper.html()).contains(item.name);
  expect(wrapper.html()).contains(item.gender);
  expect(wrapper.html()).contains(item.location.name);
  expect(wrapper.html()).toMatchSnapshot();
});
