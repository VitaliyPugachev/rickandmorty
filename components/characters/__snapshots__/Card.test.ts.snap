// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card 'Rick Sanchez' 1`] = `
"<ucard class="character-card">
  <header class="character-card__header">
    <ubadge color="blue" label="Rick Sanchez"></ubadge>
  </header>
  <section class="character-card__info">
    <div class="character-card__info-row"><span class="character-card__label"> Gender: </span><span class="character-card__value">Male</span></div>
    <div class="character-card__info-row"><span class="character-card__label"> Type: </span><span class="character-card__value"></span></div>
    <div class="character-card__info-row"><span class="character-card__label"> Location: </span><span class="character-card__value">Citadel of Ricks</span></div>
    <div class="character-card__info-row"><span class="character-card__label"> Status: </span><span class="character-card__value">Alive</span></div>
  </section>
</ucard>"
`;

exports[`Card 'Summer Smith' 1`] = `
"<ucard class="character-card">
  <header class="character-card__header">
    <ubadge color="blue" label="Summer Smith"></ubadge>
  </header>
  <section class="character-card__info">
    <div class="character-card__info-row"><span class="character-card__label"> Gender: </span><span class="character-card__value">Female</span></div>
    <div class="character-card__info-row"><span class="character-card__label"> Type: </span><span class="character-card__value">type_test</span></div>
    <div class="character-card__info-row"><span class="character-card__label"> Location: </span><span class="character-card__value">Earth</span></div>
    <div class="character-card__info-row"><span class="character-card__label"> Status: </span><span class="character-card__value">Dead</span></div>
  </section>
</ucard>"
`;
